### DISCUSSION TITLE
RFC: CausalML v1.0 roadmap

### DISCUSSION THREAD

**Discussion by jeongyoonlee** [uber/causalml#938] [Ideas]
CausalML (0.17.0) is mid-transition to sklearn-idiomatic estimators (`BaseEstimator` meta-learners, save/load, native polars/pandas). We'd like to define the road to **v1.0** and get community input before locking it in.

**What v1.0 is:** a *stabilize + expand* release positioning CausalML as the **production-scale** library for uplift / CATE estimation. It will ship through **intermediate milestones** (each released as a 0.x minor); 1.0 GA is the API freeze at the end — not a single big-bang.

**Where we see CausalML in the ecosystem:** strongest on uplift + policy optimization (Qini/AUUC/RATE, multi-treatment, value optimization) with a clean meta-learner API and SHAP interpretability. Current gaps include: no distributed/out-of-core compute, no standard benchmark-dataset loaders or PEHE/policy-risk metrics, a fragmented neural-net story, and heavy serial bootstrap-CI paths.

## Release schedule

Quarterly. Dates are targets rather than commitments; the ordering is fixed.

| Release | Target | `fit()` argument order |
|---|---|---|
| 0.18.0 | September 2026 | `fit(X, treatment, y, ...)` — deprecation warning added |
| 0.19.0 | December 2026 | unchanged, warning continues |
| 0.20.0 | March 2027 | unchanged, warning continues |
| **1.0.0** | **June 2027** | `fit(X, y, treatment, ...)` — the flip, shim removed |

## M1 — Foundation & stability
- sklearn conformance (`check_estimator` for all learners); move `black` out of runtime deps
- serialization hygiene + a pickle-size regression guard in CI
- slim the bootstrap-ensemble memory footprint
- **`fit()` argument order (#854) — decided, in progress:** adopt the scikit-learn convention `fit(X, y, treatment, ...)` (the one intentional breaking change). Existing `fit(X, treatment, y, ...)` calls keep working through the transition behind a deprecation shim that emits a `FutureWarning`.
- **unify the tree base — done** (epic #945): `UpliftTreeClassifier` / `UpliftRandomForestClassifier` now build on the shared `_tree/` Cython kernel that `CausalTreeRegressor` / `CausalRandomForestRegressor` use, via a new `UpliftClassificationCriterion`; the duplicated node / builder / splitter / bagging code in `uplift.pyx` (~2.3 KLoC) is deleted and the public classes are now thin subclasses of the kernel, with legacy signatures / defaults / predict-semantics preserved. This unblocks the `check_estimator` and serialization-hygiene items above.

## M2 — Scale-out (production headline)
- profile + optimize the estimators' own hot paths (uplift/causal-tree Cython, per-group meta-learner fitting)
- parallelize bootstrap-CI paths and respect the active joblib backend, so Ray works as a drop-in via `register_ray()` (optional `causalml[ray]` extra)
- out-of-core batch scoring of fitted learners on large populations (Ray Data) — the real large-scale win
- GPU passthrough for boosting nuisances and NN batching

## M3 — Benchmarks & evaluation
- sklearn-style `fetch_*` loaders (cached) for standard datasets: IHDP, Lalonde, Twins, Jobs, ACIC, Criteo-uplift, Hillstrom (subject to licensing)
- ground-truth metrics: PEHE / √PEHE, ε_ATE, policy risk — complementing AUUC/Qini/RATE
- a reproducible estimator × dataset leaderboard in the docs

## M4 — Modern NN & uncertainty
- consolidate the TF/PyTorch/JAX DragonNet backends onto one primary backend with friendly optional-import errors
- conformal ITE prediction intervals over any learner
- a unified uncertainty API reconciling analytic, bootstrap, and conformal intervals
- SOTA architectures: TARNet, the CATENets family (SNet/FlexTENet/OffsetNet), deep R/DR learners

## 1.0 GA
API freeze, deprecation policy + semver commitment, migration guide, and a "choosing an estimator" guide.

## Explicit non-goals for v1.0
- survival/censored-outcome HTE (low demand)
- continuous/dose-response treatments
- head-to-head DML parity with EconML (we prefer DoWhy interop)
- in-context causal foundation models (experimental, post-1.0)

## Open questions we'd love feedback on
1. Primary neural-net backend — PyTorch or JAX?
2. ~~`fit()` argument order (#854): how long a deprecation window would you need?~~ **Answered — three minor releases, ~9 months:** the `FutureWarning` ships in 0.18.0 (Sep 2026) and the flip lands in v1.0 (Jun 2027). See the schedule above and the [migration guide](https://causalml.readthedocs.io/en/latest/migration.html#fit-argument-order). Still open to hearing that this is too tight — say so below.
3. Which benchmark datasets and metrics matter most to your work?
4. Is Ray the right distributed backend, or would you prefer Dask/Spark?
5. Biggest scale/performance pain points with CausalML today?

Feedback, disagreement, and additional use cases all welcome. If there's appetite, we'll convert agreed items into tracked issues per milestone.

**Comment by jeongyoonlee** [comment]
Adding one item to M1 that I missed in the first draft: **unifying the underlying tree base for the two tree families.**

Today `CausalTreeRegressor` / `CausalRandomForestRegressor` already sit on the shared `causalml/inference/tree/_tree/` Cython kernel (`Tree`, `TreeBuilder`, `Splitter`, `RegressionCriterion`) that was vendored from sklearn 1.5.2 in #806. `UpliftTreeClassifier` / `UpliftRandomForestClassifier` don't — they live in a single 2,742-line `uplift.pyx` with their own Python-object node struct, their own recursive builder, their own bagging forest, and their own KL / ED / Chi² / DDP / IT / CIT criteria. They also don't inherit from `BaseEstimator`, which is why `save` / `load` / `_is_fitted` are monkey-patched in `tree/__init__.py` and why `check_estimator` conformance (M1) is currently blocked for these classes.

**Scope of the proposed refactor:**
- extract `BaseUpliftDecisionTree(BaseDecisionTree)` analogous to today's `BaseCausalDecisionTree`
- add `UpliftClassificationCriterion` alongside `CausalRegressionCriterion`, keeping the existing impurity math (KL, ED, Chi², DDP, IT, CIT) as pluggable criterion classes (~440 LoC preserved)
- port the builder / node / splitter usage to the shared kernel; delete the duplicated ~2.3 KLoC from `uplift.pyx`
- switch `UpliftRandomForestClassifier` from its own bagging loop to sklearn's `ForestClassifier`
- preserve public class names, arg names, and defaults; add deprecation warnings for any inevitable drift (e.g. the monkey-patched `save` / `load` becoming real methods)

**Why M1 and not M2:** it's a prerequisite for two M1 bullets already in the RFC — `check_estimator` conformance and serialization hygiene — and it changes the profiling surface for M2's "uplift/causal-tree Cython hot paths" bullet (one surface instead of two). Doing it after 1.0 GA becomes a breaking-change hazard.

**Explicit non-goals:** no change to which uplift-splitting criteria are supported; no honesty-semantics change unless separately RFC'd; no new estimator types.

**Open questions:**
1. Ship in one PR, or one PR per criterion family (KL / ED / Chi² / …)?
2. Deprecation window for the monkey-patched `save` / `load` semantics — one minor, or fold it into the M1 serialization-hygiene work?
3. Anyone relying on the current `DecisionTree` Python class as a public API surface (vs. only using it via `UpliftTreeClassifier`)? If so, we need a compatibility shim.

**Comment by ras44** [comment]
hi @jeongyoonlee I've been engaging with review requests but just realized this RFC had been posted.  Not sure if it might help to raise visibility to others.  Looks like a productive way forward 👍

**Comment by jeongyoonlee** [comment]
## Release schedule, and the answer to open question 2

Releases move to a quarterly cadence:

| Release | Target | `fit()` argument order |
|---|---|---|
| 0.18.0 | September 2026 | `fit(X, treatment, y, ...)` — deprecation warning added |
| 0.19.0 | December 2026 | unchanged, warning continues |
| 0.20.0 | March 2027 | unchanged, warning continues |
| **1.0.0** | **June 2027** | `fit(X, y, treatment, ...)` — the flip, shim removed |

Dates are targets rather than commitments. The **ordering** is the part to plan against: no signature moves before v1.0.

**Open question 2 (deprecation window length): three minor releases, roughly nine months.** The `FutureWarning` ships in 0.18.0 and the flip lands in v1.0.

Nobody asked for a longer window here, so this is a maintainer call rather than a consensus one. If nine months is too tight for your codebase, say so on this thread and it can be revisited — that is easier now than after 0.20.0.

### What to do about it

One edit, and it is safe in every version:

```python
learner.fit(X=X, treatment=treatment, y=y)   # order-independent, correct before and after v1.0
```

The migration guide covers the per-family table, the three signatures that are not a plain `treatment`/`y` swap (`IVRegressor.fit`, `BaseDRIVLearner`, and the `Sensitivity` helpers), and how to find affected calls in your own code: https://causalml.readthedocs.io/en/latest/migration.html#fit-argument-order

To list your own call sites, the warning is raised with `stacklevel=2`, so it points at your line rather than at CausalML internals:

```bash
python -W error::FutureWarning your_script.py
```

### M1 status

The argument-order work is tracked in #980. The deprecation window is closed — the shim covers every public method whose positional order changes (#975, #981), CausalML's own tests and docs are migrated to keyword calls (#982, #983), and the migration guide and changelog entry are published (#984). #985 is the flip itself and is now schedulable against the v1.0 target. #986 adds scikit-learn metadata routing, which is what actually lets a learner sit inside a `Pipeline` and closes #854.

Open questions 1, 3, 4 and 5 are still open and feedback on them is welcome.

**Comment by jeongyoonlee** [comment]
## M3 — Benchmarks & evaluation: plan, and what the licensing check turned up

Filed as #1009 – #1013. Two findings changed the shape of this from what the RFC sketched.

### The dataset decides which metric exists

The RFC lists seven datasets and three metrics as if they cross. They do not — what is computable is a property of how the data was made:

| family | datasets | ground truth | metrics that apply |
| --- | --- | --- | --- |
| simulated / semi-synthetic | `synthetic_data`, IHDP, Twins, ACIC | per-unit counterfactuals | PEHE, root PEHE, epsilon_ATE |
| experimental benchmark | LaLonde, Jobs | experimental ATE only | epsilon_ATE, policy risk |
| uplift, no ground truth | Criteo, Hillstrom | none | AUUC, Qini, RATE — already shipped |

So the leaderboard is three tables, the two uplift datasets need no new metric, and the metrics work is not blocked by the loaders: `synthetic_data` already returns `tau`, so PEHE and epsilon_ATE are testable today. #1010 starts now.

### Terms vary from MIT to nothing at all

"Subject to licensing" turns out to be the binding constraint rather than a caveat. Checking the canonical sources:

| dataset | source | terms |
| --- | --- | --- |
| LaLonde / NSW | `causaldata` (PyPI) | **MIT** |
| IHDP | `clinicalml/cfrnet` lineage, files at `fredjo.com` | **MIT** code repo |
| Twins | NBER linked birth / infant death data | **US federal, public domain** |
| ACIC | `vdorie/aciccomp` | no license file |
| Jobs | LaLonde lineage, via the Shalit et al. code | no license file |
| Criteo-Uplift | Criteo AI Lab | non-commercial, to confirm at the source |
| Hillstrom | MineThatData blog | no license file |

No license is not no permission, and these have been redistributed in the literature for years. It does mean CausalML should not vendor or mirror them. The rule the loaders will follow: **ship a loader where the source carries an explicit permissive license or is public-domain government data, fetch at runtime, never vendor.** Under that rule LaLonde, IHDP and Twins ship (#1011) and the remaining four are documented with load-it-yourself instructions (#1012), which is more useful than a loader that assumes the question away.

For Criteo and Hillstrom specifically, [`scikit-uplift`](https://github.com/maks-sh/scikit-uplift) already ships `fetch_criteo` and `fetch_hillstrom` by re-hosting both on its own S3. Pointing users there beats CausalML taking on the same redistribution.

### The issues

| | scope | status |
| --- | --- | --- |
| #1009 | fetch infrastructure — `~/causalml-data` + `CAUSALML_DATA`, SHA256, `download_if_missing`, `Bunch` / `return_X_y_t`, no network in the test suite | blocks #1011 |
| #1010 | `pehe`, `ate_error`, `policy_risk`, tested against `synthetic_data`'s `tau` | ready now |
| #1011 | `fetch_lalonde()`, `fetch_ihdp()`, `fetch_twins()` | needs #1009 |
| #1012 | `docs/datasets.rst` — availability and terms for all seven | ready now |
| #1013 | `docs/examples/benchmark_leaderboard.ipynb` — the notebook is the leaderboard, so every published number regenerates | needs #1010, #1011 |

Two design choices worth disagreeing with now rather than later. The `fetch_*` signature copies scikit-uplift's `(data_home, download_if_missing, return_X_y_t)` so code moves between the two libraries unchanged, with scikit-learn's `Bunch` return. And the loaders verify SHA256, which scikit-uplift does not — benchmark files move hosts, and a substituted file that still has a plausible length would otherwise surface as a quietly wrong benchmark number.

### Still open from the RFC

Question 3 — which datasets and metrics matter in your work — has not been answered on this thread. Rather than ask again unqualified: the plan above ships **LaLonde, IHDP and Twins** first, on licensing grounds. If ACIC or Criteo is what you actually benchmark against, say so and it changes the order.

**Reply by ras44** [reply]
@jeongyoonlee 

Responding to the open question 3 from the RFC: it seems moot.  The question asks "which datasets and metrics matter in your work" but most users are going to load their own data for their "work" and use the above-referenced datasets for tutorials or examples.  How much any of these particular datasets matter to people seems secondary to minimizing licensing risk.  The plan for shipping loaders only for MIT & public domain makes sense to me 👍

**Comment by AbdulAliMamnun** [comment]
Hi @jeongyoonlee, I'd like to contribute toward v1.0. Background: data scientist at TD Insurance working on causal inference and experimentation (uplift/CATE estimation in an insurance setting), with recent merged work in sktime (golden-output regression tests for foundation-model forecasters) and MLflow (LangChain autolog fix). I saw that #985/#986 are scheduled for June 2027 and that you filed scoped starter issues for a previous volunteer, so rather than claim something on the roadmap, I'd ask: is there a slice of M1 you'd want done now? A few things I could take on independently if useful: (a) a repeatability/regression test suite for the meta-learners and metrics, pinning outputs under fixed seeds so v1.0's API flip can't silently change estimates; (b) a scikit-learn API-compliance audit (check_estimator-style) of the learners ahead of the signature flip; (c) docs or examples for the insurance/finance uplift use case. Happy to be pointed at whatever is most useful.
